### PR TITLE
WLFY-10048 - Deploying of malformed MDB should fail

### PR DESCRIPTION
**Pull request by rpelisse**
Upstream Issue: https://issues.jboss.org/browse/WFLY-10048
Issue: https://issues.jboss.org/browse/JBEAP-14418

The fix is pretty simple, but I have extracted the logic of MDB Validation into a separate class to allow to unit test (without adding a full blown integration test, but just test the logic in a unit fashion)

### CODE REVIEW

File: ejb3/src/main/java/org/jboss/as/ejb3/util/MdbValidationsUtil.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+            return false;
+        }
+        // bean class must be public, must *not* be abstract or final
+        if (!Modifier.isPublic(flags) || Modifier.isAbstract(flags) || Modifier.isFinal(flags)) {
+            EjbLogger.DEPLOYMENT_LOGGER.mdbClassMustBePublicNonAbstractNonFinal(className);
+            return false;
+        }
+        // bean class can not have final method
+        List<String> finalMethods = new ArrayList<>(0);
+        for (MethodInfo method : mdbClass.methods()) {
```
Comment (stuartwdouglas) [inline ejb3/src/main/java/org/jboss/as/ejb3/util/MdbValidationsUtil.java]: This is not actually what the spec says (also as ClassInfo does not cover inheritance this check is not correct anyway):

5.6.4 Message Listener Method
The message-driven bean class must define the message listener methods. The signature of a message listener method must follow these rules:
The method must be declared as public.
The method must not be declared as final or static.

5.6.5 Message-Driven Bean with No-Methods Listener Interface
The following additional requirements apply for a message-driven bean with a no-methods interface:
• All non-static public methods of the bean class and of any superclasses except
java.lang.Object are exposed as message listener methods.
• Only private methods of the bean class and any superclasses except java.lang.Object may be declared final.

So basically this validation is stricter than what the spec says and can cause valid deployments to fail. It also does not check that the methods are public, and does not take superclass methods into account.

File: ejb3/src/test/java/org/jboss/as/ejb3/util/MdbValidationsUtilTest.java [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+/**
+ *
+ * @author <a href="mailto:romain@redhat.com">Romain Pelisse</a>
+ */
+public class MdbValidationsUtilTest {
+
+    @Test
+    public void mdbWithFinalClass() throws IOException {
+        MdbValidationsUtil.assertMDBClassValidity(buildClassInfoForClass("org.jboss.as.ejb3.util.InvalidMdbFinalClass"));
```
Comment (stuartwdouglas) [inline ejb3/src/test/java/org/jboss/as/ejb3/util/MdbValidationsUtilTest.java]: Why not use the type safe InvalidMdbFinalClass.class.getName() rather than a hard coded string?

File: ejb3/src/test/java/org/jboss/as/ejb3/util/MdbValidationsUtilTest.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    }
+
+    @Test
+    public void test() throws IOException {
+        MdbValidationsUtil.assertMDBClassValidity(buildClassInfoForClass("org.jboss.as.ejb3.util.InvalidMdbFinalMethods"));
+    }
+
+    private ClassInfo buildClassInfoForClass(String mdbClassName) {
+        String mdbClassNameAsResource = mdbClassName.replaceAll("\\.", "/").concat(".class");
+        Index index = indexStream(getClass().getClassLoader().getResourceAsStream(mdbClassNameAsResource)).complete();
```
Comment (stuartwdouglas) [inline ejb3/src/test/java/org/jboss/as/ejb3/util/MdbValidationsUtilTest.java]: This stream is not closed

**Review by stuartwdouglas — CHANGES_REQUESTED** [review]

**Comment by rpelisse** [comment]
@stuartwdouglas Many thanks for the review! I think I've cover most of your concerns, but please, if you have time, I would certainly appreciate an other review.

Also, please don't merge it, I'm still tracking down a failing test in the suite. I've pushed those extra changes only to allow more review on it.

**Comment by rpelisse** [comment]
OK, I've finished the change but I think they go against [AS7-1380](https://issues.jboss.org/browse/AS7-1380), so I'm discussing this with the report on https://issues.jboss.org/browse/JBEAP-14418.

**Comment by stuartwdouglas** [comment]
IMHO we only really want to report on this if it a problem that will affect the operation of the MDB. If the MDB still works fine even though it is not 100% spec compliant then I don't think we should bother.

This sort of issue generally has marginal value, as if we restrict behaviour that was previously allowed there is a good chance we will break peoples deployments. The only change we should be making here is making sure that if an MDB is currently broken the error message is clear.

**Review by rhusar — CHANGES_REQUESTED** [review]
Just rebase/squash the merge commit.

**Comment by rpelisse** [comment]
Superseeded by [PR11149](https://github.com/wildfly/wildfly/pull/11149)
